### PR TITLE
Add: new socketcan stress test

### PR DESCRIPTION
**Pull request by stanley31huang**
add new socketcan stress test (100000 loop echo test)

please refer the test results in Jira CQT-3562 issue
https://warthogs.atlassian.net/jira/software/projects/CQT/boards/418?selectedIssue=CQT-3562

Side-loading the checkbox and ran the socketcan-full test plan, please have the test results from C3:
- https://certification.canonical.com/hardware/202304-31451/submission/344073/
- https://certification.canonical.com/hardware/202304-31451/submission/344326/test-results/
following bus-off test failed as expected, I did not short the CAN_L and CAN_H in purpose.
- [after-suspend-ce-oem-socketcan/bus_off_recovery_can0](https://certification.canonical.com/hardware/202304-31451/submission/344073/test/204991/result/37757702/)
- [after-suspend-ce-oem-socketcan/bus_off_recovery_can1](https://certification.canonical.com/hardware/202304-31451/submission/344073/test/204992/result/37757703/)

### CODE REVIEW

File: checkbox-provider-ce-oem/units/socketcan/jobs.pxu [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    as to return the predicted packet.
+category_id: com.canonical.certification::socketcan
+plugin: shell
+user: root
+estimated_duration: 10m
+flags: also-after-suspend
+imports: from com.canonical.plainbox import manifest
+requires:
+    manifest.socket_can_echo_server_running == 'True'
+    socket_can_fd == 'True'
```
Comment (rickwu666666) [inline checkbox-provider-ce-oem/units/socketcan/jobs.pxu:18]: This should be `manifest.socket_can_fd`.

**Review by rickwu666666 — CHANGES_REQUESTED** [review]
Please see my in-line comment.

File: checkbox-provider-ce-oem/units/socketcan/manifest.pxu [hunk 2]
```diff
@@ -0,0 +1,5 @@
+unit: manifest entry
+id: socket_can_fd
+_prompt: Is the CAN FD capable device:
+_name: CAN FD capable devices
+value-type: bool
```
Comment (baconYao) [inline checkbox-provider-ce-oem/units/socketcan/manifest.pxu:5]: - Please leave a blank at the end of file.
- Please rename the `socket_can_fd` to be `has_socket_can_fd`

**Review by baconYao — CHANGES_REQUESTED** [review]
Please fix, thanks

File: checkbox-provider-ce-oem/units/socketcan/jobs.pxu (hunk 1 above)
Comment (stanley31huang) [reply checkbox-provider-ce-oem/units/socketcan/jobs.pxu]: thanks for the comments, fixed in latest changes.

File: checkbox-provider-ce-oem/units/socketcan/manifest.pxu (hunk 2 above)
Comment (stanley31huang) [reply checkbox-provider-ce-oem/units/socketcan/manifest.pxu]: thanks for the comments, fixed in latest changes.

**Review by rickwu666666 — APPROVED** [review]
LGTM

**Review by baconYao — CHANGES_REQUESTED** [review]
Please see my inline comments

File: checkbox-provider-ce-oem/bin/socketcan_test.py [hunk 3]
```diff
@@ -0,0 +1,325 @@
...
+        elapsed_time = end_time - start_time
+        logging.info(
+            "received %s frames in %s", len(recv_records), elapsed_time)
+
+        if len(recv_records) == count:
+            logging.info("# Checking the data in received frames")
+            failed_count = 0
+            for index, data in enumerate(original_records):
+                # validate data field in CAN packet only
+                if can_socket.destruct_packet(recv_records[index]) != \
```
Comment (baconYao) [inline checkbox-provider-ce-oem/bin/socketcan_test.py:223]: I think we can add a helper function to compare the sent and received data like below.

``` python
def is_same_packet(sent_packet, received_packet):
    s_data = can_socket.destruct_packet(sent_packet)
    r_data = can_socket.destruct_packet(received_packet)
    is_same_data = s_data == r_data
    if not is_same_data:
        logging.error("Received packet is not same as Sent packet")
        logging.error("Received packet: %s", s_data)
        logging.error("Sent data: %s", r_data)
    return is_same_data
```

So we can modify the following section
```python
if is_same_packet(data, recv_records[index]):
    failed_count += 1
```

And we can also replace the comparison at Line 143

Comment (baconYao) [reply checkbox-provider-ce-oem/bin/socketcan_test.py:223]: It would help us to implement the unit test in the future

**Review by baconYao — CHANGES_REQUESTED** [review]

File: checkbox-provider-ce-oem/bin/socketcan_test.py (hunk 3 above)
Comment (stanley31huang) [reply checkbox-provider-ce-oem/bin/socketcan_test.py]: fixed in [0b39c1d](https://github.com/canonical/checkbox-provider-ce-oem/pull/62/commits/0b39c1d9d9c8e6a6ecc9574bfc6f47fe60cfd039)

File: checkbox-provider-ce-oem/bin/socketcan_test.py [hunk 4]
```diff
@@ -0,0 +1,325 @@
...
+            raise SystemExit("CAN ECHO Stress test failed")
+
+
+def register_arguments():
+    parser = argparse.ArgumentParser(
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+        description='SocketCAN Tests')
+    parser.add_argument(
+        "-d", "--dev",
+        required=True
```
[the remark this answers is not in the document]
Comment (stanley31huang) [reply checkbox-provider-ce-oem/bin/socketcan_test.py]: fixed in [0b39c1d](https://github.com/canonical/checkbox-provider-ce-oem/pull/62/commits/0b39c1d9d9c8e6a6ecc9574bfc6f47fe60cfd039)

File: checkbox-provider-ce-oem/bin/socketcan_socket.py [hunk 5]
```diff
@@ -0,0 +1,374 @@
...
+        except OSError as e:
+            logging.error(e)
+            if e.errno == 90:
+                raise SystemExit(
+                    'ERROR: interface does not support FD Mode')
+            else:
+                raise SystemExit('ERROR: OSError on attempt to receive')
+
+
+class CANLinkState(Enum):
```
[the remark this answers is not in the document]
Comment (stanley31huang) [reply checkbox-provider-ce-oem/bin/socketcan_socket.py:205]: fixed in [0b39c1d](https://github.com/canonical/checkbox-provider-ce-oem/pull/62/commits/0b39c1d9d9c8e6a6ecc9574bfc6f47fe60cfd039)

File: checkbox-provider-ce-oem/bin/socketcan_socket.py [hunk 6]
```diff
@@ -0,0 +1,374 @@
...
+        """
+        data_struct = self.CANFD_MTU if self._fdmode else self.CAN_MTU
+        try:
+            if timeout:
+                self.sock.settimeout(timeout)
+            can_pkt = self.sock.recv(data_struct)
+            self.sock.settimeout(None)
+            return can_pkt
+        except TimeoutError:
+            logging.error("Failed to receive within %ss", 5)
```
[the remark this answers is not in the document]
Comment (stanley31huang) [reply checkbox-provider-ce-oem/bin/socketcan_socket.py]: fixed in [0b39c1d](https://github.com/canonical/checkbox-provider-ce-oem/pull/62/commits/0b39c1d9d9c8e6a6ecc9574bfc6f47fe60cfd039)

File: checkbox-provider-ce-oem/bin/socketcan_socket.py [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+    def send(self, can_packet, timeout=None):
+        """
+        Send CAN frame data through CANSocket
+
+        Args:
+            can_packet (): CAN data packet
+            timeout:
+
+        Raises:
+            SystemExit: if any error occurs during receiving CAN frame
```
[the remark this answers is not in the document]
Comment (stanley31huang) [reply checkbox-provider-ce-oem/bin/socketcan_socket.py]: fixed in [0b39c1d](https://github.com/canonical/checkbox-provider-ce-oem/pull/62/commits/0b39c1d9d9c8e6a6ecc9574bfc6f47fe60cfd039)

File: checkbox-provider-ce-oem/bin/socketcan_socket.py [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+
+        Args:
+            can_id (int):   CAN ID
+            data (byte):    CAN data packet
+            id_flag (int):  CAN ID flag
+            fd_flag (int):  additional FD flag
+            fd_frame (bol): FD frame data
+
+        Raises:
+            SystemExit: if any error occurs during receiving CAN frame
```
[the remark this answers is not in the document]
Comment (stanley31huang) [reply checkbox-provider-ce-oem/bin/socketcan_socket.py]: fixed in [0b39c1d](https://github.com/canonical/checkbox-provider-ce-oem/pull/62/commits/0b39c1d9d9c8e6a6ecc9574bfc6f47fe60cfd039)

File: checkbox-provider-ce-oem/units/socketcan/test-plan.pxu [hunk 9]
```diff
@@ -0,0 +1,50 @@
+id: ce-oem-socketcan-full
+unit: test plan
+_name: SocketCAN remote test
+_description: RS485 remote tests for devices
```
[the remark this answers is not in the document]
Comment (stanley31huang) [reply checkbox-provider-ce-oem/units/socketcan/test-plan.pxu]: fixed in [0b39c1d](https://github.com/canonical/checkbox-provider-ce-oem/pull/62/commits/0b39c1d9d9c8e6a6ecc9574bfc6f47fe60cfd039)

**Review by baconYao — APPROVED** [review]
LGTM

**Pull request merged**